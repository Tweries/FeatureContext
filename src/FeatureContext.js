import { createContext, useContext } from 'react';

const FeatureContext = createContext();

export default FeatureContext;

export function setFeatures(features) {
  if (Array.isArray(features)) {
    return {
      active: feature => features.indexOf(feature) !== -1
    };
  }
  throw new Error('the "features" argument must be an array');
}

export function useFeature() {
  return useContext(FeatureContext);
}
