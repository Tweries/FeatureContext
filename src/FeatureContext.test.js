import { render } from '@testing-library/react';
import React from 'react';
import FeatureContext, { setFeatures, useFeature } from './FeatureContext';

describe('FeatureContext', () => {
  const DANGER_V1 = 'DANGER_V1';
  const SUCCESS_V1 = 'SUCCESS_V1';
  const UNKNOWN_V1 = 'UNKNOWN_V1';

  function Message() {
    const feature = useFeature();
    if (feature.active(DANGER_V1)) {
      return <span>Oh Noes!</span>;
    }
    if (feature.active(SUCCESS_V1)) {
      return <span>YATTA</span>;
    }
    return (
      <span aria-label="man-shrugging" role="img">
        ü§∑‚Äç‚ôÇ
      </span>
    );
  }

  const scenarios = [
    {
      description: `${DANGER_V1} active`,
      feature: setFeatures([DANGER_V1])
    },
    {
      description: `${SUCCESS_V1} active`,
      feature: setFeatures([SUCCESS_V1])
    },
    {
      description: `${UNKNOWN_V1} active`,
      feature: setFeatures([UNKNOWN_V1])
    },
    {
      description: 'empty array',
      feature: setFeatures([])
    }
  ];

  scenarios.forEach(({ description, feature }) => {
    it(description, () => {
      const { container } = render(
        <FeatureContext.Provider value={feature}>
          <Message />
        </FeatureContext.Provider>
      );

      expect(container).toMatchSnapshot();
    });
  });

  it('when "null" then it should throw', () => {
    expect(() => {
      render(
        <FeatureContext.Provider value={setFeatures(null)}>
          <Message />
        </FeatureContext.Provider>
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('when "undefined" then it should throw', () => {
    expect(() => {
      render(
        <FeatureContext.Provider value={setFeatures()}>
          <Message />
        </FeatureContext.Provider>
      );
    }).toThrowErrorMatchingSnapshot();
  });
});
