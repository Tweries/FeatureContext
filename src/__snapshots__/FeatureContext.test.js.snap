// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureContext DANGER_V1 active 1`] = `
<div>
  <span>
    Oh Noes!
  </span>
</div>
`;

exports[`FeatureContext SUCCESS_V1 active 1`] = `
<div>
  <span>
    YATTA
  </span>
</div>
`;

exports[`FeatureContext UNKNOWN_V1 active 1`] = `
<div>
  <span
    aria-label="man-shrugging"
    role="img"
  >
    ü§∑‚Äç‚ôÇ
  </span>
</div>
`;

exports[`FeatureContext empty array 1`] = `
<div>
  <span
    aria-label="man-shrugging"
    role="img"
  >
    ü§∑‚Äç‚ôÇ
  </span>
</div>
`;

exports[`FeatureContext when "null" then it should throw 1`] = `"the \\"features\\" argument must be an array"`;

exports[`FeatureContext when "undefined" then it should throw 1`] = `"the \\"features\\" argument must be an array"`;
